tools/wake-format/README.md

```diff
@@ -0,0 +1,481 @@
+
+## wake-format Fluent API/DSL
+
+`wake-format` relies heavily on a custom fluent API/DSL to generate the formatting rules. Usage of the API is documented below. See [here](TODO) for design decisions, invariants, and engineering tradeoffs considered.
+
+### Usage
+
+All formatting rules start by creating a basic formatting object using `fmt()`. From there functions are added to describe the expectations of the current [CST node](TODO) and to return the appropiate text.
+
+```c++
+auto basic_fmt = fmt()
+    .space()
+    .newline();
+```
+
+creates a formatter that returns a space and a newline.
+
+Once a formatter has been defined, the `format()` function is used to run the formatter and return the formated source. The source is returned as a [wcl::doc](TODO) for efficient concatenation and geometry lookup. `format()` takes a `ctx_t` and a `CSTElement`. The `ctx_t` (context) tracks the context (such as nest level) that all subnodes should be formatted in, and the `CSTElement` is the current node being formatted.
+
+```c++
+ctx_t ctx;
+CSTElement node = ""; // An empty node
+
+wcl::doc doc = fmt()
+  .space()
+  .newline()
+  .format(ctx, node);
+
+std::string s = doc.as_string(); // " \n"
+```
+
+`format()` expects to consume the entire `CSTElement` node and will `assert()` if that is not the case. Asserting may seem heavily handed, however this is a very powerful method for identifying and fixing nodes with unexpected contents. 
+
+### API Functions & Examples
+
+**CSTElement Manipulation**
+
+Manipulation of the current state of the formatter relating to CSTElement nodes.
+
+**`token(uint8_t id)`**
+
+`assert()`s the current node.id() is `id` then emits the string representation of the node.
+
+```c++
+CSTElement node = "def";
+auto doc = fmt()
+  .token(TOKEN_KW_DEF)
+  .space()
+  .format(ctx, node);
+doc.as_string(); // "def "
+```
+
+**`token(uint8_t id, const char* str)`**
+
+`assert()`s the current node.id() is `id` then emits `str`.
+
+```c++
+CSTElement node = "def";
+auto doc = fmt()
+  .token(TOKEN_KW_DEF, "foo")
+  .space()
+  .format(ctx, node);
+doc.as_string(); // "foo "
+```
+
+**`next()`**
+
+Moves the current node forward by one element.
+
+```c++
+CSTElement node = "def id = 5";
+auto doc = fmt() // -> def
+  .next()        // -> " "
+  .next()        // -> id
+  .next()        // -> " "
+  .next()        // -> =
+  .next()        // -> " "
+  .next();       // 5
+```
+
+**Control Flow**
+
+Conditional formatting based on the state at a given point in the formatter.
+
+**`fmt_if(Predicate, Formatter)`**
+
+Applies `Formatter` if `Predicate` returns `true`. See [Predicate](#predicate) for more Predicate examples.
+
+```c++
+CSTElement node = "def id = 5";
+auto doc = fmt()
+  .fmt_if(TOKEN_KW_DEF, fmt().token(TOKEN_KW_DEF, "foo"))
+  .fmt_if(TOKEN_COMMENT, fmt().token(TOKEN_COMMENT, "bar"))
+  .format(ctx, node);
+doc.as_string(); // "foo"
+```
```

Comment:
When we do `doc.as_string()` and get `foo`, is that in replacement for the entierity of `id = 5`, or just `id`, or all of `def id = 5`?

Reply:
Well this example would actually crash because it didn't consume all the nodes.  Ignoring that this would only output "foo" because only the `def` was consumed.

Reply:
There's a partial analogy for this being dual to a parser/regex if that's helpful for you. `format` implicitly has an ending anchor at the end which is why this would fail. But ignoring that anchor, this would only match and consume the `def` and would output `foo` in its place.

Reply:
Yeah I was actually debating how to handle the fact that most of these examples actually crash. Fully consuming the input makes the example tedious, and giving a partial node is unclear.

maybe an imaginary `consume_rest()` function

Reply:
walk is also annoying to deal with in examples like this but unavoidable I think